From 25869452b925a50db6a7b9cf18104ee2092a9218 Mon Sep 17 00:00:00 2001
From: Wei Qin <qinwei61@huawei.com>
Date: Sat, 31 Jan 2026 15:59:22 +0800
Subject: [PATCH 1/1] ub: udma: Fix a bug related to create sq

driver inclusion
category: bugfix
bugzilla: https://atomgit.com/openeuler/umdk/issues/38
CVE: NA

-------------------------------------------------------

This patch fix a bug related to create sq.

Signed-off-by: Wei Qin <qinwei61@huawei.com>
---
 src/urma/hw/udma/udma_u_common.h | 1 +
 src/urma/hw/udma/udma_u_jfs.c    | 3 ++-
 src/urma/hw/udma/udma_u_jfs.h    | 7 +++++++
 3 files changed, 10 insertions(+), 1 deletion(-)

diff --git a/src/urma/hw/udma/udma_u_common.h b/src/urma/hw/udma/udma_u_common.h
index 7a2287b..c168022 100644
--- a/src/urma/hw/udma/udma_u_common.h
+++ b/src/urma/hw/udma/udma_u_common.h
@@ -38,6 +38,7 @@
 
 #define UDMA_JFC_DB_OFFSET 0
 #define UDMA_MIN(x, y) ((x) < (y) ? (x) : (y))
+#define UDMA_MAX(x, y) ((x) > (y) ? (x) : (y))
 #define RTE_SET_USED(x) (void)(x)
 
 struct udma_u_doorbell {
diff --git a/src/urma/hw/udma/udma_u_jfs.c b/src/urma/hw/udma/udma_u_jfs.c
index 85aa5f1..5e7fcd7 100644
--- a/src/urma/hw/udma/udma_u_jfs.c
+++ b/src/urma/hw/udma/udma_u_jfs.c
@@ -129,7 +129,8 @@ int udma_u_create_sq(struct udma_u_jetty_queue *sq, urma_jfs_cfg_t *cfg)
 	udma_u_init_sq_param(sq, cfg);
 
 	sqe_bb_cnt = sq_cal_wqebb_num(SQE_WRITE_NOTIFY_CTL_LEN,
-				      cfg->max_sge, UDMA_JFS_WQEBB);
+				      get_max_sge_num(cfg->max_sge, sq->max_inline_size),
+				      UDMA_JFS_WQEBB);
 	if (sqe_bb_cnt > MAX_SQE_BB_NUM)
 		sqe_bb_cnt = MAX_SQE_BB_NUM;
 	sq->sqe_bb_cnt = sqe_bb_cnt;
diff --git a/src/urma/hw/udma/udma_u_jfs.h b/src/urma/hw/udma/udma_u_jfs.h
index d0389d0..a069545 100644
--- a/src/urma/hw/udma/udma_u_jfs.h
+++ b/src/urma/hw/udma/udma_u_jfs.h
@@ -139,6 +139,13 @@ static inline uint32_t sq_cal_wqebb_num(uint32_t sqe_ctl_len, uint32_t sge_num,
 		wqebb_size + (uint32_t)1;
 }
 
+static inline uint32_t get_max_sge_num(uint8_t max_sge, uint32_t max_inline_size)
+{
+	uint32_t size = (max_inline_size == 0) ?
+			1 : ((max_inline_size - (uint32_t)1) / UDMA_SGE_SIZE + (uint32_t)1);
+	return UDMA_MAX(max_sge, size);
+}
+
 urma_status_t udma_u_post_sq_wr(struct udma_u_context *udma_ctx,
 				struct udma_u_jetty_queue *sq, urma_jfs_wr_t *wr,
 				urma_jfs_wr_t **bad_wr);
-- 
2.34.1

